src/xrpld/app/tx/detail/LoanSet.cpp

```diff
@@ -0,0 +1,675 @@
+        return temINVALID;
+
+    // Copied from preflight2
+    if (counterPartySig)
+    {
+        if (auto const ret = ripple::detail::preflightCheckSimulateKeys(
+                ctx.flags, *counterPartySig, ctx.j))
+            return *ret;
+    }
+
+    if (auto const brokerID = ctx.tx[~sfLoanBrokerID];
+        brokerID && *brokerID == beast::zero)
+        return temINVALID;
+
+    return tesSUCCESS;
+}
+
+NotTEC
+LoanSet::checkSign(PreclaimContext const& ctx)
+{
+    if (auto ret = Transactor::checkSign(ctx))
+        return ret;
+
+    // Counter signer is optional. If it's not specified, it's assumed to be
+    // `LoanBroker.Owner`. Note that we have not checked whether the
+    // loanbroker exists at this point.
+    auto const counterSigner = [&]() -> std::optional<AccountID> {
+        if (auto const c = ctx.tx.at(~sfCounterparty))
+            return c;
+
+        if (auto const broker =
+                ctx.view.read(keylet::loanbroker(ctx.tx[sfLoanBrokerID])))
+            return broker->at(sfOwner);
+        return std::nullopt;
+    }();
+    if (!counterSigner)
+        return temBAD_SIGNER;
+
+    // Counterparty signature is optional. Presence is checked in preflight.
+    if (!ctx.tx.isFieldPresent(sfCounterpartySignature))
+        return tesSUCCESS;
+    auto const counterSig = ctx.tx.getFieldObject(sfCounterpartySignature);
+    return Transactor::checkSign(
+        ctx.view,
+        ctx.flags,
+        ctx.parentBatchId,
+        *counterSigner,
+        counterSig,
+        ctx.j);
+}
+
+XRPAmount
+LoanSet::calculateBaseFee(ReadView const& view, STTx const& tx)
+{
+    auto const normalCost = Transactor::calculateBaseFee(view, tx);
+
+    // Compute the additional cost of each signature in the
+    // CounterpartySignature, whether a single signature or a multisignature
+    XRPAmount const baseFee = view.fees().base;
+
+    // Counterparty signature is optional, but getFieldObject will return an
+    // empty object if it's not present.
+    auto const counterSig = tx.getFieldObject(sfCounterpartySignature);
+    // Each signer adds one more baseFee to the minimum required fee
+    // for the transaction. Note that unlike the base class, the single signer
+    // is counted if present. It will only be absent in a batch inner
+    // transaction.
+    std::size_t const signerCount = counterSig.isFieldPresent(sfSigners)
+        ? counterSig.getFieldArray(sfSigners).size()
+        : (counterSig.isFieldPresent(sfTxnSignature) ? 1 : 0);
+
+    return normalCost + (signerCount * baseFee);
+}
+
+std::vector<OptionaledField<STNumber>> const&
+LoanSet::getValueFields()
+{
+    static std::vector<OptionaledField<STNumber>> const valueFields{
+        ~sfPrincipalRequested,
+        ~sfLoanOriginationFee,
+        ~sfLoanServiceFee,
+        ~sfLatePaymentFee,
+        ~sfClosePaymentFee
+        // Overpayment fee is really a rate. Don't check it here.
+    };
+
+    return valueFields;
+}
+
+std::uint32_t
```

Comment:
It should be `static` function.

Reply:
> It should be `static` function.

I assume you're talking about `getStartDate`, in which case, this is fixed in 1ceff3dde70b877fdfc30a3ab2322a5f13b56070